plugin/src/main/java/hudson/plugins/swarm/PluginImpl.java

```diff
@@ -221,6 +227,38 @@ public void doCreateSlave(StaplerRequest req, StaplerResponse rsp, @QueryParamet
         }
     }
 
+    /**
+     * Allows the client to mark a slave offline
+     */
+    public void doMarkSlaveOffline(StaplerRequest req, StaplerResponse rsp, @QueryParameter String name,
+                              @QueryParameter String reason, @QueryParameter String secret) throws IOException {
+        if (!getSwarmSecret().equals(secret)) {
+            rsp.setStatus(SC_FORBIDDEN);
+            return;
+        }
+
+        Jenkins jenkins = Jenkins.getInstance();
+
+        jenkins.checkPermission(SlaveComputer.DISCONNECT);
+
+        Node node = getNodeByName(name, rsp);
+
+        node.toComputer().setTemporarilyOffline(true, new OfflineCause.UserCause(User.current(), reason));
+
+        normalResponse(req, rsp, "<response/>"); //TODO: not sure what to send here
+    }
+
+    public void doGetSlaveReadyForShutdown(StaplerRequest req, StaplerResponse rsp, @QueryParameter String name) throws IOException {
+        Jenkins jenkins = Jenkins.getInstance();
```

Comment:
Regarding checking for secrets/permissions: I can't see any disadvantages to checking, and I can't see any advantages to not checking. In contrast, if we don't check, we increase the attack surface for potential security issues. Based on the above reasoning, I think I'd prefer to check secrets/permissions.